### PR TITLE
[RPC] direct connect

### PR DESCRIPTION
**Pull request by rgrinberg**
I moved the connection finding logic to the RPC library. This makes it possible
for clients to establish the connection directly after calling `Where.get ()`
and converting this value into a chan.

As a result, I had to move the socket from the build directory to
`Path.Source.root`. The reason being that the build directory can be customized
with --build-dir, and hence we cannot rely on the socket to be in `_build`

### CODE REVIEW

**Comment by cwong-ocaml** [comment]
> The reason being that the build directory can be customized
> with `--build-dir`, and hence we cannot rely on the socket to be in `_build`

Could we pass this directory to the client instead? I'm a bit uncomfortable with having an internal socket file being in the source directory. I recognize that there isn't a good way to get this directory if the server is started externally, but I'm not convinced that this is a use case that we should try to tailor a perfect experience around.

**Comment by rgrinberg** [comment]
> Could we pass this directory to the client instead?

We could.

> but I'm not convinced that this is a use case that we should try to tailor a perfect experience around.

Note that this use case is partially motivated by @jeremiedimino mentioning that you'd like to experiment with running the build directory elsewhere for performance.

**Comment by cwong-ocaml** [comment]
> Note that this use case is partially motivated by @jeremiedimino mentioning that you'd like to experiment with running the build directory elsewhere for performance.

For our case, I think that this is not an unknown but just a *different* build directory, so being able to pass it to the client library would be enough for us.

To be clear, the use case I don't think is worth supporting is in which the server is started by an unknown entity with an unknown build directory; in such a case I think a best-effort "look in `_build`, fail if it doesn't exist" is probably for the best.

**Comment by jeremiedimino** [comment]
To be fair, for Jane Street it is fine is the socket is not in the workspace. If we decide to put the build directory somewhere else, we can just hard-code the location of the build directory in our Emacs plugin.

**Comment by rgrinberg** [comment]
Okay, so it is settled then. I will change the PR to try and discover the build directory, but leave a function for the user to pass the build directory as a parameter if `--build-dir` is used.

**Comment by jeremiedimino** [comment]
BTW, you could rename the socket do have "dune" in the filename, even if it is inside the build directory. This way, there is less chance a client accidentally connecting to a non-Dune server.

**Comment by rgrinberg** [comment]
Okay, this is should be good to go. I reverted the location of the socket back to the build directory. The new api requires you to pass the build directory explicitly. I get the feeling that trying to discover the build directory for the user isn't such a good idea. We should encourage the user to think about `--build-dir`, rather than relying on `_build` somewhere up in the tree.

**Comment by rgrinberg** [comment]
@cwong-ocaml this is ready for review then

**Review by cwong-ocaml — APPROVED** [review]
Generally looks good, mostly stylistic concerns.

Sorry about the delayed review; Github seemed to destroy my session every time a new batch of commits was pushed.

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 1]
```diff
@@ -314,4 +314,42 @@ module V1 : sig
...
+
+    module type S = sig
+      type 'a fiber
+
+      val get : build_dir:string -> t option fiber
+
+      val default : build_dir:string -> t
+    end
+
+    module Make (Fiber : sig
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:332]: Things are indeed inconsistent but I'm not sure about your suggestion. We currently have to two functors:

* `Where.Make` to create a channel from an address
* `Client` to create a client from a channel

How about we change `Client` to `Client.Make`. That would make things consistent.

[the remark this answers is not in the document]
Comment (cwong-ocaml) [reply otherlibs/dune-rpc/dune_rpc.mli:332]: `Client.Make` is good.

[the remark this answers is not in the document]
Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:334]: Renamed.

**Pull request merged**